addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoDeviceProtocol.java

```diff
@@ -0,0 +1,125 @@
+     * @param roomKey the non-empty room key
+     * @param deviceKey the non-empty device key
+     * @throws IOException Signals that an I/O exception has occurred.
+     */
+    public NeeoDeviceProtocol(NeeoHandlerCallback callback, String roomKey, String deviceKey) throws IOException {
+        Objects.requireNonNull(callback, "callback cannot be null");
+        Objects.requireNonNull(roomKey, "roomKey cannot be empty");
+        Objects.requireNonNull(deviceKey, "deviceKey cannot be empty");
+
+        this.callback = callback;
+
+        final NeeoBrainApi api = callback.getApi();
+        if (api == null) {
+            throw new IllegalArgumentException("NeeoBrainApi cannot be null");
+        }
+
+        neeoRoom = api.getRoom(roomKey);
+        if (neeoRoom == null) {
+            throw new IllegalArgumentException("Room (" + roomKey + ") was not found in the NEEO Brain");
+        }
+
+        neeoDevice = neeoRoom.getDevices().getDevice(deviceKey);
+        if (neeoDevice == null) {
+            throw new IllegalArgumentException(
+                    "Device (" + deviceKey + ") was not found in the NEEO Brain for room (" + roomKey + ")");
+        }
+    }
+
+    /**
+     * Returns the callback being used
+     *
+     * @return the non-null callback
+     */
+    public NeeoHandlerCallback getCallback() {
+        return callback;
+    }
+
+    /**
+     * Refresh the macro status.
+     *
+     * @param macroKey the non-null macro key
+     */
+    public void refreshMacroStatus(String macroKey) {
+        NeeoUtil.requireNotEmpty(macroKey, "macroKey cannot be empty");
+
+        final NeeoMacro macro = neeoDevice.getMacros().getMacro(macroKey);
+        if (macro != null) {
+            callback.stateChanged(UidUtils.createChannelId(NeeoConstants.DEVICE_GROUP_MACROSID, macroKey),
+                    OnOffType.OFF);
+        }
+    }
+
+    /**
+     * Sets the macro status. If the status is true, the macro will be triggered. If false, nothing occurs
+     *
+     * @param macroKey the non-null macro key
+     * @param start whether to start (true) or stop (false) the recipe
+     */
+    public void setMacroStatus(String macroKey, boolean start) {
+        NeeoUtil.requireNotEmpty(macroKey, "macroKey cannot be empty");
+
+        final NeeoBrainApi api = callback.getApi();
+        if (api == null) {
+            logger.debug("API is null [likely bridge is offline]");
+        } else {
+            try {
+                if (start) {
+                    api.triggerMacro(neeoRoom.getKey(), neeoDevice.getKey(), macroKey);
+                    callback.scheduleTask(() -> {
+                        refreshMacroStatus(macroKey);
+                    }, 500);
+                }
+            } catch (IOException e) {
+                logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
+                // callback.statusChanged(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```

Comment:
if you cannot communicate with your device, why not setting it to `OFFLINE`?

Reply:
As you can see - originally I did.  However, realize we just got to firmware .43 and there are still issues on their side that cause random errors.  Because of that, I don't want to mark it offline when we just hit a bug in their firmware.  When we get to an official 1.0 release - let's revisit this.  If you'd like, I can either leave it as is or maybe document that?

Reply:
Please document it somewhere, so you do not forget to remove this hack.